### ISSUE TITLE
V5 Why is Sanitation or Deny-list not recommended

### ISSUE THREAD

**Issue by timurozkul** [OWASP/ASVS#2381]
Sanitization or/and Deny-listing should be added as supplementary recommendations.

Started the discussion on OWASP Cheatsheets however due lack of input and ASVS also not recommending a deny-list or sanitization it would worth discussing it here also. 

Summarizing the discussion https://github.com/OWASP/CheatSheetSeries/issues/1541:
All other recommendation on input validation require individual implementation making it high workload and more error prone vs implementing a sanitization library or deny-listing as middleware which is quick win. Sanitization/Deny-listing libraries help fend off a degree of attacks. 

I have not found one valid literature that explains why input sanitization / deny-listing is bad other then it gives a false sense of security and that it can strip out legitimate characters. Can someone explain why filtering out or blocking `; DROP TABLE` from an input field for `"favorite food"` would be considered a bad?

I was asked to provide good use cases: All WAF providers utilize this method And why is html sanitization an exception? 

I suggest we put sanitization / deny-listing back on ASVS as supplementary recommendation unless there is a solid argument against it.

**Comment by elarlang** [comment]
TLDR: I don't agree with this direction or proposal.

Few things that come to mind:
* The false sense of security, as you mentioned yourself - I have met the attitude "we have defended by enterprise grade WAF, we are safe"
* Often bypassing those rules are as trivial as messing with lower-case vs upper-case letter, using different whitespace characters or encodings
* I consider it not as a defense, but as a "slowdowner" - you need to fix underlying problems in your application anyway
* Automatically blocking or modifying user-input is against integrity and it must be synchronized with business logic rules - can you do it or not. User input must reach to the application as it was written.
* Deny-list is "list of bad things you know" and you can make the application secure against yourself. In case an attacker knows more bad things than you, attacker can bypass it. 

> Can someone explain why filtering out or blocking ; DROP TABLE from an input field for "favorite food" would be considered a bad?

But why you need to use deny-list instead of allow-list filtering? And even if you use sanitization, it does not replace the need to defend against SQL injection.

In short - you can consider this as a extra layer of "defense" to slow down attackers, but this is not actual defense that an application needs.

> I suggest we put sanitization / deny-listing back on ASVS as supplementary recommendation unless there is a solid argument against it.

Put it back? I rechecked and at least from ASVS v3.0.1 it has been only allow-list for input validation and for sanitization, there no no "put back".

So, it's a "no" from me.

**Comment by tghosth** [comment]
1) I agree with everything @elarlang wrote. We are looking to focus on leading practices in ASVS and positive allowlists for input validation are part of that. 
2) Semantically aware sanitization for certain dangerous sinks is mentioned in V5.2 so I think that is probably sufficient.

**Comment by elarlang** [comment]
If this was the only reason you read out from here, I recommend investigating the meaning of data integrity.

**Comment by elarlang** [comment]
With arguments there is one tiny problem -  you ignore them :)

If explanations from different people and from different repositories is not enough for you, we can call it religion. For this kind of situation, I have standard recommendations:
* Please recommend this defense for example for stack overflow
* Give a talk at a security conference about that and see what kind of feedback you'll get
